### PR TITLE
[Enhancement] Multiple barcode per Item

### PR DESCRIPTION
**Pull request by GSLabIt**
- ability to add multiple barcode for an Item. 
- ability to validate UPC-A and EAN if barcode type is selected.

Some GIFs:

UPC-A Validation:
rydiwxs2hz

Find duplicates:
rnesyble9u

Search on POS:
m1hftrccys

Search on Stock Entry:
frh7f1h0my

### CODE REVIEW

**Comment by joezsweet** [comment]
https://github.com/frappe/erpnext/issues/6809

File: erpnext/patches.txt [hunk 1]
```diff
@@ -479,3 +479,4 @@ erpnext.patches.v9_2.rename_net_weight_in_item_master
 erpnext.patches.v9_2.delete_process_payroll
 erpnext.patches.v10_0.add_agriculture_domain
 erpnext.patches.v10_0.add_non_profit_domain
+erpnext.patches.v10_0item_barcode_childtable_migrate
```
Comment (tundebabzy) [inline erpnext/patches.txt]: > erpnext.patches.v10_0item_barcode_childtable_migrate

You forgot the '.'

Comment (tundebabzy) [inline erpnext/patches.txt]: > erpnext.patches.v10_0item_barcode_childtable_migrate

You forgot the '.'

Comment (joezsweet) [reply erpnext/patches.txt]: wooppps ..

Comment (tundebabzy) [inline erpnext/patches.txt]: > erpnext.patches.v10_0item_barcode_childtable_migrate

You forgot the '.'

Comment (tundebabzy) [inline erpnext/patches.txt]: > erpnext.patches.v10_0item_barcode_childtable_migrate

You forgot the '.'

**Review by tundebabzy — CHANGES_REQUESTED** [review]

File: erpnext/patches.txt (hunk 1 above)
Comment (joezsweet) [reply erpnext/patches.txt]: fixed

**Comment by joezsweet** [comment]
Checked Travis error, it seems to be not related to the PR

**Comment by Mohd200** [comment]
Every Barcode should have its own UOM (For POS).
Is that Applicable in this scenario ??

**Comment by joezsweet** [comment]
Nope, actually it's only used to store multiple barcode for same item, not per UOM.

**Comment by skarthik3** [comment]
Hi joezsweet 
Great job. I just want to make a small suggestion on this instead of changing so many files why not use customize form change bar code field "data type" to "small text" store multiple barcodes as comma seperated values and add bar code to search fields. Assuming barcode Code 128 has max 48 characters we can store good number of bar codes.

**Comment by joezsweet** [comment]
@skarthik3 barcode needs to be unique and would be possible to validate based on barcode type; beside that you'll need to change files anyway ..

**Comment by olamide2** [comment]
@joezsweet 

Any progress on this? 

It is a fantastic feature we are really waiting on

**Comment by joezsweet** [comment]
@olamide2 waiting to be merged ..

**Comment by olamide2** [comment]
@tundebabzy anyway to accelerate this ? I have a need for this urgently. 

Regards

**Review by tundebabzy — CHANGES_REQUESTED** [review]
```(python)
Executing erpnext.patches.v10_0.item_barcode_childtable_migrate in dev1 (3b42553c4e3241e8)
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/utils/bench_helper.py", line 94, in <module>
    main()
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/utils/bench_helper.py", line 18, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/tunde/bench-repo/dev/env/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/commands/__init__.py", line 24, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/commands/site.py", line 222, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/migrate.py", line 31, in migrate
    frappe.modules.patch_handler.run_all()
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/modules/patch_handler.py", line 29, in run_all
    if not run_single(patchmodule = patch):
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/modules/patch_handler.py", line 63, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/modules/patch_handler.py", line 83, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + ".execute")()
  File "/home/tunde/bench-repo/dev/apps/erpnext/erpnext/patches/v10_0/item_barcode_childtable_migrate.py", line 14, in execute
    doc.append("barcodes", {"barcode": item.get("barcode")})
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/model/base_document.py", line 138, in append
    value = self._init_child(value, key)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/model/base_document.py", line 164, in _init_child
    value["doctype"] = self.get_table_field_doctype(key)
  File "/home/tunde/bench-repo/dev/apps/frappe/frappe/model/base_document.py", line 278, in get_table_field_doctype
    return self.meta.get_field(fieldname).options
AttributeError: 'NoneType' object has no attribute 'options'
```

File: requirements.txt [hunk 2]
```diff
@@ -2,3 +2,4 @@ frappe
 unidecode
 pygithub
 googlemaps
+python-stdnum
```
Comment (tundebabzy) [inline requirements.txt]: `python-stdnum` is not used anywhere in the code

**Review by tundebabzy — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 2 above)
Comment (joezsweet) [reply requirements.txt]: have a look at 

https://github.com/joezsweet/erpnext/blob/c95675c074408e6c522499960d910495784e3b7e/erpnext/stock/doctype/item/item.py#L486-L496

Comment (joezsweet) [reply requirements.txt]: it seems Item Barcode Doctype is missing ..

Comment (joezsweet) [reply requirements.txt]: have a look at 

https://github.com/joezsweet/erpnext/tree/multi-barcode/erpnext/stock/doctype/item_barcode

Comment (tundebabzy) [reply requirements.txt]: ok

**Comment by tundebabzy** [comment]
Still fails with the same error. The problem stems from the fact that in the patch, the retrieved Item doc doesn't have the `barcodes` field. Reload the doc

**Comment by joezsweet** [comment]
sorry, my bad ..fixed

File: erpnext/patches/v10_0/item_barcode_childtable_migrate.py [hunk 3]
```diff
@@ -9,7 +9,6 @@
 def execute():
     items_barcode = frappe.db.sql("SELECT name, barcode FROM tabItem WHERE barcode IS NOT NULL", as_dict=1)
     frappe.reload_doc("stock", "doctype", "item")
```
Comment (tundebabzy) [inline erpnext/patches/v10_0/item_barcode_childtable_migrate.py]: Also reload `Item Barcode`

**Review by tundebabzy — CHANGES_REQUESTED** [review]

**Comment by joezsweet** [comment]
ohh ..sorry  ..tested on mine with both table in there :( ..fixed

File: erpnext/patches/v10_0/item_barcode_childtable_migrate.py (hunk 3 above)
Comment (joezsweet) [reply erpnext/patches/v10_0/item_barcode_childtable_migrate.py]: ohh ..sorry ..tested on mine with both table in there :( ..fixed

**Comment by tundebabzy** [comment]
@joezsweet 
Seems like validation is not working properly. Correct me if I'm wrong
screenshot-2018-1-11 non-iron blue twill slim fit button cuff shirt-l - non-iron blue twill slim fit button cuff shirt-l

**Comment by joezsweet** [comment]
@tundebabzy it seems the code u're testing is a wrong one: 

schermata 2018-01-11 alle 22 51 26

i'm using python-stdnum to verify validity

**Review by tundebabzy — APPROVED** [review]
Looks good.

**Comment by mfazilp** [comment]
Waiting for this feature

**Comment by olamide2** [comment]
Really waiting ..as in " should have had it seen yesterday " waiting

**Comment by olamide2** [comment]
@tundebabzy ...thank you for the good work...anything else stopping this from been merged ? I really need this urgently

**Comment by olamide2** [comment]
When can we merger this ? Anything else holding it back ??

**Comment by mfazilp** [comment]
Waiting for this feature

**Comment by olamide2** [comment]
Waiting still. ..

**Comment by bkm888** [comment]
While the concept looks good, not all PR make it into the core system, even if the code is sound. They have to be approved by the devs controlling the code to make sure it fits with their roadmap for the project. Barcode management has it's own set of pitfalls they are probably considering. I know I had trouble getting anyone to think about barcodes this time last year. Hopefully your luck is better.
bkm@jmibarcodes.com

**Comment by olamide2** [comment]
Thanks @bkm888

But barcode is an integral part of inventory management. Most companies will not even consider an ERP solution without  a standard barcode mgt system.

I hope the devs do understand this part.

**Comment by bkm888** [comment]
@olamide2 
I agree that barcoding is important, and ERPNext has been using them for a long time now quite successfully in POS, stock moves, and several other applications. I do also agree that having the ability to assign more than one barcode to an item is very useful. However, it will likely require a major change in how they are saved in the database which may have a ripple effect into other database functions. There is likely a great deal of regression testing to be done to make sure nothing else is affected. When a change like this one is requested, it is only expanding on an already working function, yet it entails making database changes that need to be tested thoroughly. When a function like barcoding already works and you only want to expand it, it's priority may not be the same for the developers as it is to the person requesting it. Patience is key here.

**Comment by joezsweet** [comment]
@bkm888 
" However, it will likely require a major change in how they are saved in the database which may have a ripple effect into other database functions. There is likely a great deal of regression testing to be done to make sure nothing else is affected." 

All queries has been tested, actually only few were affected by the change ..

"This line in particular is the one I see as possibly the most troubling:ability to validate UPC-A and EAN if barcode type is selected."

That's an option, you can set to check or not ..i think it's useful when you're entering barcode manually ..btw feel free to not use it ..if u not select barcode type no validation is performed ..

File: erpnext/controllers/accounts_controller.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 from erpnext.accounts.party import get_party_account_currency, validate_party_frozen_disabled
 from erpnext.exceptions import InvalidCurrency
 
-force_item_fields = ("item_group", "barcode", "brand", "stock_uom")
+force_item_fields = ("item_group", "barcodes", "brand", "stock_uom")
```
Comment (rohitwaghchaure) [inline erpnext/controllers/accounts_controller.py]: What is use of barcodes here? because here barcodes is the child table name

Comment (joezsweet) [reply erpnext/controllers/accounts_controller.py]: you're right it can be removed

File: erpnext/stock/doctype/item/item.py [hunk 5]
```diff
@@ -2,744 +2,777 @@
...
-		self.validate_attributes()
-		self.validate_variant_attributes()
-		self.validate_website_image()
-		self.make_thumbnail()
-		self.validate_fixed_asset()
-		self.validate_retain_sample()
-
-		if not self.get("__islocal"):
-			self.old_item_group = frappe.db.get_value(self.doctype, self.name, "item_group")
-			self.old_website_item_groups = frappe.db.sql_list("""select item_group
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/item/item.py]: Why spaces over tabs? In standard, we use tabs for indentation. Kindly don't change the format.

Comment (joezsweet) [reply erpnext/stock/doctype/item/item.py]: that's was a pep8 standard thought

File: erpnext/stock/doctype/item/item.py [hunk 6]
```diff
@@ -282,8 +296,8 @@ def set_variant_context(self, context):
 			# load variants
 			# also used in set_attribute_context
 			context.variants = frappe.get_all("Item",
-				filters={"variant_of": self.name, "show_variant_in_website": 1},
-				order_by="name asc")
+											  filters={"variant_of": self.name, "show_variant_in_website": 1},
+											  order_by="name asc")
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/item/item.py]: Kindly remove the spaces!

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 7]
```diff
@@ -2,69 +2,81 @@
...
+        'batch_no_data': get_batch_no_data(),
+        'barcode_data': get_barcode_data(items_list),
+        'tax_data': get_item_tax_data(),
+        'price_list_data': get_price_list_data(doc.selling_price_list),
+        'bin_data': get_bin_data(pos_profile),
+        'pricing_rules': get_pricing_rule_data(doc),
+        'print_template': print_template,
+        'pos_profile': pos_profile,
+        'meta': get_meta()
+    }
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Kindly remove the spaces!

File: erpnext/stock/doctype/item/item.py (hunk 5 above)
Comment (rohitwaghchaure) [reply erpnext/stock/doctype/item/item.py]: Oh, we like to use tabs not the spaces!
https://www.youtube.com/watch?v=SsoOG6ZeyUI

**Comment by joezsweet** [comment]
@rohitwaghchaure should be ok now ..

File: erpnext/stock/doctype/item/item.py (hunk 6 above)
Comment (joezsweet) [reply erpnext/stock/doctype/item/item.py]: @rohitwaghchaure done ..

**Comment by albertusgeyser** [comment]
Possible?

#12855

**Comment by rmehta** [comment]
@rohitwaghchaure is this good to go?

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 8]
```diff
@@ -51,20 +60,23 @@ def get_pos_data():
...
 def get_meta():
 	doctype_meta = {
 		'customer': frappe.get_meta('Customer'),
 		'invoice': frappe.get_meta('Sales Invoice')
 	}
 
-	for row in frappe.get_all('DocField', fields = ['fieldname', 'options'],
-		filters = {'parent': 'Sales Invoice', 'fieldtype': 'Table'}):
+	for row in frappe.get_all('DocField', fields=['fieldname', 'options'],
+							  filters={'parent': 'Sales Invoice', 'fieldtype': 'Table'}):
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 9]
```diff
@@ -123,35 +138,39 @@ def update_multi_mode_option(doc, pos_profile):
 
 		doc.append('payments', payment_mode)
 
+
 def get_mode_of_payment(doc):
-	return frappe.db.sql(""" select mpa.default_account, mpa.parent, mp.type as type from `tabMode of Payment Account` mpa,
-		 `tabMode of Payment` mp where mpa.parent = mp.name and mpa.company = %(company)s""", {'company': doc.company}, as_dict=1)
+	return frappe.db.sql(""" select mpa.default_account, mpa.parent, mp.type as type from `tabMode of Payment Account` mpa, \
+						`tabMode of Payment` mp where mpa.parent = mp.name and mpa.company = %(company)s""", {'company': doc.company}, as_dict=1)
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 10]
```diff
@@ -248,9 +274,29 @@ def get_batch_no_data():
...
+			itemwise_barcode[item.item_code].append(barcode)
+
+	return itemwise_barcode
+
+
 def get_item_tax_data():
-	# get default tax of an item
-	# example: {'Consulting Services': {'Excise 12 - TS': '12.000'}}
+		# get default tax of an item
+		# example: {'Consulting Services': {'Excise 12 - TS': '12.000'}}
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 11]
```diff
@@ -262,17 +308,19 @@ def get_item_tax_data():
 
 	return itemwise_tax
 
+
 def get_price_list_data(selling_price_list):
 	itemwise_price_list = {}
 	price_lists = frappe.db.sql("""Select ifnull(price_list_rate, 0) as price_list_rate,
 		item_code from `tabItem Price` ip where price_list = %(price_list)s""",
-		{'price_list': selling_price_list}, as_dict=1)
+								{'price_list': selling_price_list}, as_dict=1)
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 12]
```diff
@@ -297,9 +346,10 @@ def get_pricing_rule_data(doc):
 						and ifnull(company, '') in (%(company)s, '') and disable = 0 and %(date)s
 						between ifnull(valid_from, '2000-01-01') and ifnull(valid_upto, '2500-12-31')
 						order by priority desc, name desc""",
-						{'company': doc.company, 'price_list': doc.selling_price_list, 'date': nowdate()}, as_dict=1)
+									  {'company': doc.company, 'price_list': doc.selling_price_list, 'date': nowdate()}, as_dict=1)
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 13]
```diff
@@ -338,14 +388,16 @@ def make_invoice(doc_list={}, email_queue_list={}, customers_list={}):
...
 def validate_records(doc):
 	validate_item(doc)
 
+
 def get_customer_id(doc, customer=None):
 	cust_id = None
 	if doc.get('customer_pos_id'):
 		cust_id = frappe.db.get_value('Customer',
-			{'customer_pos_id': doc.get('customer_pos_id')}, 'name')
+									  {'customer_pos_id': doc.get('customer_pos_id')}, 'name')
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 14]
```diff
@@ -380,28 +434,31 @@ def add_customer(data):
...
+										'territory') or _('All Territories')
+
 
 def get_customer_group(data):
 	if data.get('customer_group'):
 		return data.get('customer_group')
 
 	return frappe.db.get_single_value('Selling Settings',
-		'customer_group') or frappe.db.get_value('Customer Group', {'is_group': 0}, 'name')
+										'customer_group') or frappe.db.get_value('Customer Group', {'is_group': 0}, 'name')
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/accounts/doctype/sales_invoice/pos.py [hunk 15]
```diff
@@ -380,28 +434,31 @@ def add_customer(data):
...
 
-def make_contact(args,customer):
+
+def make_contact(args, customer):
 	if args.get('email_id') or args.get('phone'):
 		name = frappe.db.get_value('Dynamic Link',
-			{'link_doctype': 'Customer', 'link_name': customer, 'parenttype': 'Contact'}, 'parent')
+									{'link_doctype': 'Customer', 'link_name': customer, 'parenttype': 'Contact'}, 'parent')
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/pos.py]: Extra spaces

File: erpnext/stock/doctype/item/item.py [hunk 16]
```diff
@@ -2,30 +2,41 @@
...
-from frappe.utils import (cstr, flt, cint, getdate, now_datetime, formatdate,
-	strip, get_timestamp, random_string)
+import json
+import erpnext
+import frappe
+from erpnext.controllers.item_variant import (ItemVariantExistsError,
+												copy_attributes_to_variant,
+												get_variant,
+												make_variant_item_code,
+												validate_item_variant_attributes)
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/item/item.py]: Extra spaces

File: erpnext/stock/doctype/item/item.py [hunk 17]
```diff
@@ -147,19 +158,19 @@ def set_opening_stock(self):
 
 		# default warehouse, or Stores
 		default_warehouse = (self.default_warehouse
-			or frappe.db.get_single_value('Stock Settings', 'default_warehouse')
-			or frappe.db.get_value('Warehouse', {'warehouse_name': _('Stores')}))
+							 or frappe.db.get_single_value('Stock Settings', 'default_warehouse')
+							 or frappe.db.get_value('Warehouse', {'warehouse_name': _('Stores')}))
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/item/item.py]: Extra spaces

File: erpnext/stock/doctype/item/item.py [hunk 18]
```diff
@@ -147,19 +158,19 @@ def set_opening_stock(self):
...
-				qty=self.opening_stock, rate=self.valuation_rate)
+										   qty=self.opening_stock, rate=self.valuation_rate)
 
 			stock_entry.add_comment("Comment", _("Opening Stock"))
 
 	def make_route(self):
 		if not self.route:
 			return cstr(frappe.db.get_value('Item Group', self.item_group,
-				'route')) + '/' + self.scrub((self.item_name if self.item_name else self.item_code) + '-' + random_string(5))
+											'route')) + '/' + self.scrub((self.item_name if self.item_name else self.item_code) + '-' + random_string(5))
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/item/item.py]: Extra spaces

**Comment by rohitwaghchaure** [comment]
@joezsweet 

Kindly check pos.py and item.py file again, Saw extra spaces in this file, kindly remove that extra spaces

**Comment by nabinhait** [comment]
@rohitwaghchaure If extra spaces is the only problem, remove those spaces and merge yourself.

**Comment by joezsweet** [comment]
@rohitwaghchaure done, should be ok now ..sorry for that

**Comment by nabinhait** [comment]
@joezsweet Please fix the merge conflict

**Comment by joezsweet** [comment]
@nabinhait not sure how to fix json conflict, could u help?

**Comment by nabinhait** [comment]
@joezsweet ok, let me fix the conflict and merge

**Pull request merged**